// @devrev/typescript-sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Workflow = exports.client = exports.publicSDK = exports.workflowSDK = exports.internalSDK = exports.betaSDK = void 0;
exports.betaSDK = __importStar(require("./auto-generated/beta/beta-devrev-sdk"));
exports.internalSDK = __importStar(require("./auto-generated/internal/private-internal-devrev-sdk"));
exports.workflowSDK = __importStar(require("./auto-generated/public-beta/public-beta-devrev-sdk"));
exports.publicSDK = __importStar(require("./auto-generated/public-devrev-sdk"));
var client_1 = require("./client_setup/client");
Object.defineProperty(exports, "client", { enumerable: true, get: function () { return client_1.client; } });
var workflow_1 = require("./workflow/workflow");
Object.defineProperty(exports, "Workflow", { enumerable: true, get: function () { return workflow_1.Workflow; } });
__exportStar(require("./snap-ins/types"), exports);
