// @devrev/typescript-sdk/dist/logger.js
"use strict";
/* eslint-disable no-console */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLogger = exports.ConsoleLogger = exports.LogLevel = void 0;
/**
 * Severity levels for log entries
 */
var LogLevel;
(function (LogLevel) {
    LogLevel["ERROR"] = "error";
    LogLevel["WARN"] = "warn";
    LogLevel["INFO"] = "info";
    LogLevel["DEBUG"] = "debug";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
/**
 * Default logger which logs to stdout and stderr
 */
class ConsoleLogger {
    constructor() {
        this.level = LogLevel.INFO;
        this.name = '';
    }
    getLevel() {
        return this.level;
    }
    /**
     * Sets the instance's log level so that only messages which are equal or more severe are output to the console.
     */
    setLevel(level) {
        this.level = level;
    }
    /**
     * Set the instance's name, which will appear on each log line before the message.
     */
    setName(name) {
        this.name = name;
    }
    /**
     * Log a debug message
     */
    debug(...msg) {
        if (ConsoleLogger.isMoreOrEqualSevere(LogLevel.DEBUG, this.level)) {
            const label = ConsoleLogger.labels.get(LogLevel.DEBUG);
            const color = ConsoleLogger.colors[LogLevel.DEBUG];
            console.debug(color + label, this.name, ...msg, '\x1b[0m'); // Reset color
        }
    }
    /**
     * Log an info message
     */
    info(...msg) {
        if (ConsoleLogger.isMoreOrEqualSevere(LogLevel.INFO, this.level)) {
            const label = ConsoleLogger.labels.get(LogLevel.INFO);
            const color = ConsoleLogger.colors[LogLevel.INFO];
            console.info(color + label, this.name, ...msg, '\x1b[0m'); // Reset color
        }
    }
    /**
     * Log a warning message
     */
    warn(...msg) {
        if (ConsoleLogger.isMoreOrEqualSevere(LogLevel.WARN, this.level)) {
            const label = ConsoleLogger.labels.get(LogLevel.WARN);
            const color = ConsoleLogger.colors[LogLevel.WARN];
            console.warn(color + label, this.name, ...msg, '\x1b[0m'); // Reset color
        }
    }
    /**
     * Log an error message
     */
    error(...msg) {
        if (ConsoleLogger.isMoreOrEqualSevere(LogLevel.ERROR, this.level)) {
            const label = ConsoleLogger.labels.get(LogLevel.ERROR);
            const color = ConsoleLogger.colors[LogLevel.ERROR];
            console.error(color + label, this.name, ...msg, '\x1b[0m'); // Reset color
        }
    }
    /**
     * Helper to compare two log levels and determine if a is equal or more severe than b
     */
    static isMoreOrEqualSevere(a, b) {
        return ConsoleLogger.severity[a] >= ConsoleLogger.severity[b];
    }
}
exports.ConsoleLogger = ConsoleLogger;
/** Map of labels for each log level */
ConsoleLogger.labels = (() => {
    const entries = Object.entries(LogLevel);
    const map = entries.map(([key, value]) => [value, `[${key}] `]);
    return new Map(map);
})();
/** Map of ANSI color codes for each log level */
ConsoleLogger.colors = {
    [LogLevel.ERROR]: '\x1b[31m',
    [LogLevel.WARN]: '\x1b[33m',
    [LogLevel.INFO]: '\x1b[36m',
    [LogLevel.DEBUG]: '\x1b[35m', // Magenta color
};
/** Map of severity as comparable numbers for each log level */
ConsoleLogger.severity = {
    [LogLevel.ERROR]: 400,
    [LogLevel.WARN]: 300,
    [LogLevel.INFO]: 200,
    [LogLevel.DEBUG]: 100,
};
let instanceCount = 0;
/**
 * INTERNAL interface for getting or creating a named Logger.
 */
function getLogger(name, level, existingLogger) {
    // Get a unique ID for the logger.
    const instanceId = instanceCount;
    instanceCount += 1;
    // Set up the logger.
    const logger = (() => {
        if (existingLogger !== undefined) {
            return existingLogger;
        }
        return new ConsoleLogger();
    })();
    logger.setName(`[devrev:${name}:${instanceId}]`);
    if (level !== undefined) {
        logger.setLevel(level);
    }
    return logger;
}
exports.getLogger = getLogger;
